File: microservice-architecture/microservice-basics/ts/src/InventoryClient.ts
/**
 * @file InventoryClient.ts - 库存服务契约接口
 *
 * 【设计模式】
 *   - 依赖倒置原则（DIP）：OrderService 依赖此接口而非具体实现。
 *   - 策略模式（Strategy Pattern）：不同的实现类代表不同的调用策略。
 *
 * 【架构思想】
 *   TypeScript 的 interface 在编译后会被擦除，但在开发时提供了类型安全保障。
 *   服务间通过契约接口解耦，调用方不关心被调服务是本地还是远程。
 *
 * 【开源对比】
 *   - gRPC：通过 .proto 文件自动生成 TypeScript 接口
 *   - tRPC：端到端类型安全的 RPC 框架
 */

/** 库存服务契约接口 */
export interface InventoryClient {
  /**
   * 预留库存
   * @param sku      商品 SKU 编码
   * @param quantity 预留数量
   * @returns true=预留成功，false=库存不足
   */
  reserve(sku: string, quantity: number): boolean
}

File: microservice-architecture/microservice-basics/python/src/inventory_client.py
"""
inventory_client.py - 库存服务契约接口

【设计模式】
  - 依赖倒置原则（DIP）：OrderService 依赖此接口而非具体实现，
    使得本地调用和远程调用可以通过替换实现类来切换。
  - 策略模式（Strategy Pattern）：不同的实现类代表不同的调用策略。

【架构思想】
  在微服务架构中，服务间通过契约（接口/协议）解耦。
  调用方只依赖契约，不关心被调服务是本地实现还是远程服务。

【开源对比】
  - Python 生态中通常用 ABC（Abstract Base Class）定义服务契约
  - gRPC-Python 通过 .proto 文件自动生成客户端接口
  本示例用基类 + NotImplementedError 模拟抽象接口。
"""


class InventoryClient(object):
    """库存服务契约接口。

    所有库存服务实现（本地/远程）都必须实现此接口。
    OrderService 通过此接口访问库存服务，实现解耦。
    """

    def reserve(self, sku, quantity):
        """预留库存。

        Args:
            sku: 商品 SKU 编码
            quantity: 预留数量

        Returns:
            bool: True=预留成功，False=库存不足

        Raises:
            NotImplementedError: 子类必须实现此方法
        """
        raise NotImplementedError()
